Bottom: 4bb8b1e2af578041984ee401219e821c90debd73
Top:    7cbec879d4691538acd1489b65f289c405bee96f
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-03-30 13:32:56 +0800

tun: xmit_more support

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index a82bced..c9f0a83 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -896,10 +896,12 @@ static netdev_tx_t tun_net_xmit(struct sk_buff *skb, struct net_device *dev)
 	if (skb_array_produce(&tfile->tx_array, skb))
 		goto drop;
 
-	/* Notify and wake up reader process */
-	if (tfile->flags & TUN_FASYNC)
-		kill_fasync(&tfile->fasync, SIGIO, POLL_IN);
-	tfile->socket.sk->sk_data_ready(tfile->socket.sk);
+	if (!skb->xmit_more) {
+		/* Notify and wake up reader process */
+		if (tfile->flags & TUN_FASYNC)
+			kill_fasync(&tfile->fasync, SIGIO, POLL_IN);
+		tfile->socket.sk->sk_data_ready(tfile->socket.sk);
+	}
 
 	rcu_read_unlock();
 	return NETDEV_TX_OK;
